### PR TITLE
[lang] MatrixType bug fix: support star & loop operations over matrix-typed Expr

### PR DESCRIPTION
**Pull request by jim19930609**
Issue: https://github.com/taichi-dev/taichi/issues/5819

### CODE REVIEW

File: python/taichi/lang/ast/ast_transformer.py [hunk 1]
```diff
@@ -279,6 +279,9 @@ def process_generators(ctx, node, now_comp, func, result):
             return func(ctx, node, result)
         with ctx.static_scope_guard():
             _iter = build_stmt(ctx, node.generators[now_comp].iter)
+
+        if impl.current_cfg().real_matrix and isinstance(_iter, impl.Expr):
+            _iter = [Expr(x) for x in ctx.ast_builder.expand_expr([_iter.ptr])]
```
Comment (strongoier) [inline python/taichi/lang/ast/ast_transformer.py:284]: It seems problematic if there's only one `for`.

File: python/taichi/lang/ast/ast_transformer.py [hunk 2]
```diff
@@ -30,6 +30,20 @@
     from ast import unparse
 
 
+def reshape_list(flat_list, target_shape):
+    if len(target_shape) < 2:
+        return flat_list
+
+    curr_list = []
+    dim = target_shape.pop()
```
Comment (AD1024) [inline python/taichi/lang/ast/ast_transformer.py:38]: It's internal use so maybe OK: this line will exhaust `target_shape` in a way the caller may not expect.

**Review by AD1024 — APPROVED** [review]
Looks good!

File: python/taichi/lang/ast/ast_transformer.py (hunk 2 above)
Comment (jim19930609) [reply python/taichi/lang/ast/ast_transformer.py:38]: Ah thanks for pointing it out. I should provide another entrance to make a copy of `target_shape`

**Review by strongoier — APPROVED** [review]
LGTM!

**Pull request merged**